src/db.c

```diff
@@ -332,11 +383,52 @@ robj *dbRandomKey(redisDb *db) {
     }
 }
 
+/* Return random non-empty dictionary from this DB. */
+dict *getFairRandomDict(redisDb *db) {
+    if (db->dict_count == 1) return db->dict[0];
+    unsigned long target = randomULong();
+    unsigned long long int key_count = dbSize(db);
+    if (!key_count) return db->dict[0];
+    target %= key_count; /* Random key index in range [0..KEY_COUNT). */
+    /* In linearly enumerated key space, find a slot that contains target key. */
+    dbIterator dbit;
+    dbInitIterator(&dbit, db);
+    dict *d = NULL;
+    while ((d = dbNextDict(&dbit))) {
+        unsigned long long int ds = dictSize(d);
+        if (target < ds) { /* Found dict that contains target key. */
+            return d;
+        }
+        target -= ds;
```

Comment:
Practically speaking, won't this invalidate most of the L1 cache as we are bringing in on average 8k value to do the size check. If most commands are evicting items out of the LRU, this seems really inefficient.

Reply:
Random dict code has been refactored and this should no longer be an issue.